### PR TITLE
Add metadata support

### PR DESCRIPTION
**Pull request by t4ccer**
To add metadata support we have to use patched versions of `plutus` and `plutus-apps` until IOG will manage their dependencies, and rewrite  `plutus-ledger-constraints` so as I don't have better place to describe how to attach metadata I will do it here. `Ledger.Constraints` now exports `mustIncludeMetadata` function that can be used to include `TxMetadata` (defined in `Ledger.Constraints.Metadata`) encoded as JSON object as onchain metadata.

### CODE REVIEW

File: examples/plutus-nft/app/Main.hs [hunk 1]
```diff
@@ -37,15 +33,38 @@ instance HasDefinitions MintNFTContracts where
...
-        mintNft tokenName
+        mintNft p
 
-newtype MintNFTContracts = MintNFT TokenName
+newtype MintNFTContracts = MintNFT MintParams
   deriving stock (Show)
 
 $(deriveJSON defaultOptions ''MintNFTContracts)
 
+{-
```
Comment (samuelWilliams99) [inline examples/plutus-nft/app/Main.hs:45]: Should this still be here?

File: src/BotPlutusInterface/Files.hs [hunk 2]
```diff
@@ -148,6 +158,17 @@ writeValidatorScriptFile pabConf validatorScript =
...
+-- | Writes metadata file under the given folder
+writeMetadataFile ::
+  forall (w :: Type) (effs :: [Type -> Type]).
+  Member (PABEffect w) effs =>
+  PABConfig ->
+  BuiltinByteString ->
+  Eff effs (Either (FileError ()) Text)
+writeMetadataFile pabConf metadata =
+  let filepath = metadataFilePath pabConf metadata
+   in fmap (const filepath) <$> writeFileRaw @w (Text.unpack filepath) metadata
```
Comment (samuelWilliams99) [inline src/BotPlutusInterface/Files.hs:170]: Maybe this?
```suggestion
   in (filepath <$) <$> writeFileRaw @w (Text.unpack filepath) metadata
```
Or
```suggestion
   in const filepath <<$>> writeFileRaw @w (Text.unpack filepath) metadata
```
using (<<$>>) from Cardano.Prelude

**Review by samuelWilliams99 — COMMENTED** [review]
Does this lock us to a plutus version until its upstreamed?

I'm also assuming that metadata doesn't affect fees, does it?

**Comment by samuelWilliams99** [comment]
I see, im assuming the cardano-cli min fee takes metadata into account.

**Comment by samuelWilliams99** [comment]
We're updating to plutus-apps `905ab07ca1515a2999529616ae7015616a7f6deb` for a fix attempt, so you may need to bump your custom versions to that. I'll keep you posted.

**Comment by samuelWilliams99** [comment]
Also do we have a PR to plutus-apps btw?

**Comment by t4ccer** [comment]
Not yet, to make PR to `plutut-apps` I need some changes in `plutus`, so before going through all of it I'd like to confirm that it actually works

**Comment by samuelWilliams99** [comment]
This should be good to merge once conflicts are resolved

**Pull request merged**